src/main/java/com/aws/greengrass/security/SecurityService.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.security;
+
+import com.aws.greengrass.config.CaseInsensitiveString;
+import com.aws.greengrass.logging.api.Logger;
+import com.aws.greengrass.logging.impl.LogManager;
+import com.aws.greengrass.security.exceptions.KeyLoadingException;
+import com.aws.greengrass.security.exceptions.ServiceProviderConflictException;
+import com.aws.greengrass.security.exceptions.ServiceUnavailableException;
+import lombok.AccessLevel;
+import lombok.Getter;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import javax.net.ssl.KeyManager;
+
+public class SecurityService {
+    private static final Logger logger = LogManager.getLogger(SecurityService.class);
+    private static final String KEY_TYPE = "keyType";
+    private static final String KEY_URI = "keyUri";
+
+    @Getter(AccessLevel.PACKAGE)
+    private final ConcurrentMap<CaseInsensitiveString, CryptoKeySpi> cryptoKeyProviderMap;
+
+    public SecurityService() {
+        this.cryptoKeyProviderMap = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Register crypto key provider for the key type.
+     *
+     * @param keyProvider Crypto key provider
+     * @throws ServiceProviderConflictException if key type is already registered
+     */
+    public void registerCryptoKeyProvider(CryptoKeySpi keyProvider) throws ServiceProviderConflictException {
+        CaseInsensitiveString keyType = new CaseInsensitiveString(keyProvider.supportedKeyType());
+        logger.atInfo().kv(KEY_TYPE, keyType).log("Register crypto key service provider");
+        CryptoKeySpi provider = cryptoKeyProviderMap.computeIfAbsent(keyType, k -> keyProvider);
+        if (!provider.equals(keyProvider)) {
+            logger.atError().kv(KEY_TYPE, keyType)
+                    .log("Crypto key service provider for the key type is already registered");
+            throw new ServiceProviderConflictException(String.format("Key type %s provider is registered", keyType));
+        }
+    }
+
+    /**
+     * Deregister crypto key provide for the key type.
+     *
+     * @param keyProvider Crypto key provider
+     */
+    public void deregisterCryptoKeyProvider(CryptoKeySpi keyProvider) {
```

Comment:
Is it expected to return silently if user tries to deregister previously unregistered CryptoKeyProvider? 
How about we return a `boolean` instead of  `void` here, that shows the result of deregister operation?

Reply:
It's intended to have API idempotent. Also returning false is probably confusing - whether it's unregistered or not.